### PR TITLE
[Merged by Bors] - chore: bump to nightly-2023-07-15

### PR DESCRIPTION
**Pull request by kim-em**
Various adaptations to changes when `Fin` API was moved to Std. One notable change is that many lemmas are now stated in terms of `i ≠ 0` (for `i : Fin n`) rather then `i.1 ≠ 0`, and as a consequence many `Fin.vne_of_ne` applications have been added or removed, mostly removed.

---

- [x] depends on: https://github.com/leanprover/std4/pull/188

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This will need a fix to Std4.

**Comment by semorrison** [comment]
This is currently failing in `tests/CommDiag.lean`, but I can't reproduce the failure locally.

**Comment by Vtec234** [comment]
Now it is failing on the `simpNF` linter.

File: Mathlib/Data/Fin/Basic.lean [hunk 1]
```diff
@@ -391,48 +305,18 @@ def ofNat'' [NeZero n] (i : ℕ) : Fin n :=
...
--- the name now that there is no clash.
-@[simp]
-theorem val_zero (n : ℕ) [NeZero n] : ((0 : Fin n) : ℕ) = 0 :=
-  rfl
 #align fin.coe_zero Fin.val_zero
-#align fin.val_zero' Fin.val_zero
 
 @[simp]
-theorem mk_zero [NeZero n] : (⟨0, NeZero.pos n⟩ : Fin n) = (0 : Fin _) :=
+theorem val_zero' (n : ℕ) [NeZero n] : ((0 : Fin n) : ℕ) = 0 :=
```
Comment (urkud) [inline Mathlib/Data/Fin/Basic.lean:311]: What does `Fin.val_zero` in Std say? Could you please add a docstring about it?

Comment (semorrison) [inline Mathlib/Data/Fin/Basic.lean:311]:
```suggestion
/--
The `Fin.val_zero` in `Std` only applies in `Fin (n+1)`.
This one instead uses a `NeZero n` typeclass hypothesis.
-/
theorem val_zero' (n : ℕ) [NeZero n] : ((0 : Fin n) : ℕ) = 0 :=
```

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

File: Mathlib/Data/Fin/Basic.lean [hunk 2]
```diff
@@ -391,48 +305,23 @@ def ofNat'' [NeZero n] (i : ℕ) : Fin n :=
...
+-/
 @[simp]
-theorem mk_zero [NeZero n] : (⟨0, NeZero.pos n⟩ : Fin n) = (0 : Fin _) :=
+theorem val_zero' (n : ℕ) [NeZero n] : ((0 : Fin n) : ℕ) = 0 :=
   rfl
-#align fin.mk_zero Fin.mk_zero
 
-@[simp]
-theorem zero_le [NeZero n] (a : Fin n) : 0 ≤ a :=
-  Nat.zero_le a.val
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Basic.lean:410]: Doesn't this one need to stay too?

Comment (eric-wieser) [reply Mathlib/Data/Fin/Basic.lean:410]: The one in Std4 is
```
@[simp] theorem zero_le (a : Fin (n + 1)) : 0 ≤ a := Nat.zero_le a.val
```
which doesn't work for `NeZero`

File: Mathlib/Data/Fin/Basic.lean [hunk 3]
```diff
@@ -391,48 +305,23 @@ def ofNat'' [NeZero n] (i : ℕ) : Fin n :=
...
-theorem zero_lt_one : (0 : Fin (n + 2)) < 1 :=
-  Nat.zero_lt_one
 #align fin.zero_lt_one Fin.zero_lt_one
-
-@[simp]
-theorem not_lt_zero (a : Fin n.succ) : ¬a < 0 :=
-  fun.
 #align fin.not_lt_zero Fin.not_lt_zero
-
-theorem pos_iff_ne_zero [NeZero n] (a : Fin n) : 0 < a ↔ a ≠ 0 := by
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Basic.lean:422]: idem

File: Mathlib/Data/Fin/Basic.lean [hunk 4]
```diff
@@ -935,78 +735,23 @@ theorem succ_one_eq_two' : Fin.succ (1 : Fin (n + 2)) = 2 :=
...
 #align fin.last_le_iff Fin.last_le_iff
-
-@[simp]
-theorem lt_add_one_iff {n : ℕ} {k : Fin (n + 1)} : k < k + 1 ↔ k < last n := by
-  rw [← not_iff_not]
-  simp
 #align fin.lt_add_one_iff Fin.lt_add_one_iff
-
-@[simp]
-theorem le_zero_iff {n : ℕ} [NeZero n] {k : Fin n} : k ≤ 0 ↔ k = 0 :=
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Basic.lean:990]: idem

File: Mathlib/Data/Fin/Basic.lean [hunk 5]
```diff
@@ -1231,86 +911,35 @@ theorem castIso_castSucc {n' : ℕ} {h : n + 1 = n' + 1} {i : Fin n} :
...
+#align fin.cast_succ_pos Fin.castSucc_pos'
 
-@[simp]
-theorem castSucc_eq_zero_iff [NeZero n] (a : Fin n) : castSucc a = 0 ↔ a = 0 :=
-  Fin.ext_iff.trans <| (Fin.ext_iff.trans <| by simp).symm
 #align fin.cast_succ_eq_zero_iff Fin.castSucc_eq_zero_iff
-
-theorem castSucc_ne_zero_iff [NeZero n] (a : Fin n) : castSucc a ≠ 0 ↔ a ≠ 0 :=
-  not_iff_not.mpr <| castSucc_eq_zero_iff a
 #align fin.cast_succ_ne_zero_iff Fin.castSucc_ne_zero_iff
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Basic.lean:942]: idem

bors r-

File: Mathlib/Data/Fin/Basic.lean [hunk 6]
```diff
@@ -2045,6 +1328,19 @@ end AddGroup
...
+  if i.1 < p.1 then castSucc i else succ i
+
+/-- `predAbove p i` embeds `i : Fin (n+1)` into `Fin n` by subtracting one if `p < i`. -/
+def predAbove (p : Fin n) (i : Fin (n + 1)) : Fin n :=
+  if h : castSucc p < i then i.pred ((ne_iff_vne i 0).mpr (Nat.not_eq_zero_of_lt h))
+  else i.castLT (Nat.lt_of_le_of_lt (Nat.ge_of_not_lt h) p.2)
+
+/-- `castPred` embeds `i : Fin (n + 2)` into `Fin (n + 1)`
+by lowering just `last (n + 1)` to `last n`. -/
+def castPred (i : Fin (n + 2)) : Fin (n + 1) := predAbove (last n) i
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Basic.lean:1342]: Should we put these back to how they were in mathlib3, and align them?

Comment (eric-wieser) [reply Mathlib/Data/Fin/Basic.lean:1342]: [#portdatafi](https://github.com/leanprover-community/mathlib4/blob/f4e42872269c9502498e2c2c9e5fae8753936920/Mathlib/Data/Fin/Basic.lean#L2032-L2042)

Comment (eric-wieser) [reply Mathlib/Data/Fin/Basic.lean:1342]: I think these were removed in [this diff](https://github.com/leanprover-community/mathlib4/commit/906f72214116b27baae2f0dd5d7fe6d119c6c373#diff-6ae63812d312aba302a17e36f91f095aee62f00d3cb2f4966f176ea040474db6)

File: Mathlib/Data/Fin/Basic.lean [hunk 7]
```diff
@@ -2045,6 +1328,19 @@ end AddGroup
...
 section SuccAbove
 
+/-- `succAbove p i` embeds `Fin n` into `Fin (n + 1)` with a hole around `p`. -/
+def succAbove (p : Fin (n + 1)) (i : Fin n) : Fin (n + 1) :=
+  if i.1 < p.1 then castSucc i else succ i
+
+/-- `predAbove p i` embeds `i : Fin (n+1)` into `Fin n` by subtracting one if `p < i`. -/
+def predAbove (p : Fin n) (i : Fin (n + 1)) : Fin n :=
+  if h : castSucc p < i then i.pred ((ne_iff_vne i 0).mpr (Nat.not_eq_zero_of_lt h))
+  else i.castLT (Nat.lt_of_le_of_lt (Nat.ge_of_not_lt h) p.2)
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Basic.lean:1338]: This surely belongs in the `section PredAbove`?

File: Mathlib/Data/Fin/Basic.lean (hunk 6 above)
Comment (eric-wieser) [reply Mathlib/Data/Fin/Basic.lean]: Ah, only `succAbove` was changed; I moved the other lemmas back to their original (unbundled) homes.

**Comment by eric-wieser** [comment]
bors d+

The `succAbove` thing is probably out of scope.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**